lib/cousins.ml

```diff
@@ -73,3 +73,302 @@ let br_inter_is_empty b1 b2 =
+  in
+  loop (10000, -10000) l
+
+let rec ascendants base acc l i =
+  match l with
+  | [] -> acc
+  (* TODO type for this tuple?; why list of level? *)
+  | (ip, _, _, lev :: _ll) :: l -> (
+      match get_parents (poi base ip) with
+      | None -> ascendants base acc l i
+      | Some ifam ->
+          let cpl = foi base ifam in
+          let ifath = get_father cpl in
+          let imoth = get_mother cpl in
+          let acc = [ (ifath, [], ifath, [ lev + 1 ]) ] @ acc in
+          let acc = [ (imoth, [], imoth, [ lev + 1 ]) ] @ acc in
+          ascendants base acc l i)
+  | _ :: l ->
+      !GWPARAM.syslog `LOG_WARNING
+        "Unexpected empty level list in ascend computation\n";
+      ascendants base acc l i
+
+(* descendants des ip de liste1 sauf ceux présents dans liste2 *)
+let descendants_aux base liste1 liste2 =
+  let liste2 = List.map (fun (ip, _, _, _) -> ip) liste2 in
+  let rec loop0 acc = function
+    | [] -> acc
+    | (ip, ifaml, ipar0, lev :: _ll) :: l ->
+        let fams = Array.to_list (get_family (poi base ip)) in
+        let chlds =
+          (* accumuler tous les enfants de ip *)
+          let rec loop1 acc fams =
+            (* iterer sur chaque famille *)
+            match fams with
+            | [] -> acc
+            | ifam :: fams ->
+                let children =
+                  let rec loop2 acc2 children =
+                    match children with
+                    | [] -> acc2
+                    | ipch :: children ->
+                        loop2
+                          ((ipch, ifam :: ifaml, ipar0, [ lev - 1 ]) :: acc2)
+                          children
+                  in
+                  loop2 [] (Array.to_list (get_children (foi base ifam)))
+                in
+                loop1 (acc @ children) fams
+          in
+          loop1 [] fams
+        in
+        let chlds =
+          List.fold_left (* on élimine les enfants présents dans l2 *)
+            (fun acc (ip, ifaml, ipar, lev) ->
+              if List.mem ip liste2 then acc else (ip, ifaml, ipar, lev) :: acc)
+            [] chlds
+        in
+        loop0 (chlds @ acc) l
+    | _ :: l ->
+        !GWPARAM.syslog `LOG_WARNING
+          "Unexpected empty level list in descend computation\n";
+        loop0 acc l
+  in
+  loop0 [] liste1
+
+let descendants base cousins_cnt i j =
+  let liste1 = cousins_cnt.(i).(j - 1) in
+  let liste2 = if i > 0 then cousins_cnt.(i - 1).(j - 1) else [] in
+  descendants_aux base liste1 liste2
+
+let init_cousins_cnt base max_a_l max_d_l p =
+  match (!cousins_t, !cousins_dates_t) with
+  | Some t, Some d_t -> (t, d_t)
+  | _, _ ->
+      let t', d_t' =
+        let max_l1, max_l2 = max_l1_l2_aux max_a_l max_d_l in
+        Printf.sprintf "******** Compute %d × %d table ********\n" max_l1 max_l2
+        |> !GWPARAM.syslog `LOG_WARNING;
+        (* TODO test for Sys.max_array_length *)
+        let () = load_ascends_array base in
+        let () = load_couples_array base in
```

Comment:
there is a `clear_(ascends|couples)_array` function too, should we call it once `init_cousins_cnt` is finished?